Tribes/src/app/_components/building-type/building-type.component.ts

```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
 
   addNewBuilding() {
     if (this.goldAmount > COST_NEW_BUILDING) {
+      this.goldAmount = this.goldAmount - COST_NEW_BUILDING;
       this.buildingService.addNewBuilding(this.type);
     } else {
       this.errorMessage =  'You do not have enough money.';
     }
   }
 
   checkIfBuildingIsUpgradeable(level: number): boolean {
-    return this.getNumberOfBuildingsByLevel(level) !== 0 && this.townhallLevel > level && this.goldAmount >= this.getUpgradingCost(level);
+    return this.getNumberOfBuildingsByLevel(level) !== 0
+            && this.townhallLevel > level
+            && this.goldAmount >= this.getUpgradingCost(level)
+            && !this.checkIfAllBuildingByLevelIsUnderConstruction(level);
+  }
+
+  checkUpgradingConditions(level: number): string {
+    if (this.getNumberOfBuildingsByLevel(level) === 0) {
+      return 'You don\'t have any building of level ' + level;
+    } else if (this.checkIfAllBuildingByLevelIsUnderConstruction(level)) {
+      return 'All buildings of level ' + level + ' are upgrading';
+    } else {
+      if (this.townhallLevel <= level && this.goldAmount < this.getUpgradingCost(level)) {
+        return 'You don\'t have enough money and the townhall level is too low.';
+      } else if (this.townhallLevel <= level) {
+        return 'The townhall level is too low.';
+      } else if (this.goldAmount < this.getUpgradingCost(level)) {
+        return 'You don\'t have enough money';
+      }
+    }
+  }
+
+  upgradeBuilding(level: number): void {
+    if (this.type === 'townhall' && !this.checkIfTownhallIsUpgrading()) {
+      if (this.goldAmount >= this.getUpgradingCost(level)) {
+        const idToUpgrade: number = this.buildings.find(building => building.type === 'TOWNHALL').id;
+        this.goldAmount = this.goldAmount - this.getUpgradingCost(level);
+        this.buildingService.upgradeBuilding(idToUpgrade, level + 1);
+      } else {
+        this.errorMessage = this.checkTownhallUpgradingConditions(level);
+      }
+    } else {
+      if (this.checkIfBuildingIsUpgradeable) {
+        const idToUpgrade: number = this.buildings
+                                  .find(building => building.level === level).id;
+        this.goldAmount = this.goldAmount - this.getUpgradingCost(level);
+        this.buildingService.upgradeBuilding(idToUpgrade, level + 1);
+      } else {
+        this.errorMessage = this.checkUpgradingConditions(level);
+      }
+    }
+  }
+
+  checkIfAllBuildingByLevelIsUnderConstruction(level: number): boolean {
+    const readyBuildings: Building[] = this.buildings.filter(building => !this.buildingService.checkIfBuildingIsProgressing(building));
+    return readyBuildings.length === 0;
+  }
```

Comment:
I am not sure if we have decided whether you can be upgrading multiple levels at the same time. I assume that this lets you be able to upgrade 1 per level? It is good, just did we agree on this?

Reply:
This condition check, if the user have any building of a level, which is not under construction not upgrading  at the moment. Only this case can upgrade one more building. We can have the situation, that all buildings of a level are upgrading, but they have the previous level until the process is finished, without this check we could begin the upgrading process of the same building twice.